### PR TITLE
Added support for automatic timeseries collection creation for 5.0

### PR DESCRIPTION
**Pull request by rozza**
KAFKA-228

### CODE REVIEW

File: src/main/java/com/mongodb/kafka/connect/util/TimeseriesValidation.java [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+      final MongoNamespace namespace,
+      final MongoSinkTopicConfig config) {
+
+    try {
+
+      Document collStats = getCollStats(mongoClient, namespace);
+      if (collStats.containsKey(TIMESERIES)) {
+        return;
+      }
+      if (collStats.getInteger("nindexes") > 0) {
```
Comment (stIncMale) [inline src/main/java/com/mongodb/kafka/connect/util/TimeseriesValidation.java:124]: What is achieved by the `nindexes` check? It seems excessive to me for the following reasons:

- [`collStats`](https://docs.mongodb.com/v5.0/reference/command/collStats/) documentation states: _"If the collection does not exist, collStats returns an error message"_, which I expect is represented by an exception in the Java driver. Therefore, if we are executing the `collStats.getInteger("nindexes") > 0` check, we already know that the collection identified by `namespace` exists.
- The [Technical Design: Time-series Collection](https://docs.google.com/document/d/11GqNugOwdnwgU8LyOziLdrYhHTICX3UpcUFJVvU5Lw8/edit#heading=h.i4sji0le9arb) states _"A collStats command being run against a time-series collection contains the following additional fields ... `timeseries`"_. Taking into account 1, we know that if we are executing the `collStats.getInteger("nindexes") > 0` check, the collection not only exists but also is not a time-series collection.

File: src/integrationTest/java/com/mongodb/kafka/connect/ConnectorValidationIntegrationTest.java [hunk 2]
```diff
@@ -511,6 +627,14 @@ private boolean isReplicaSetOrSharded() {
     }
   }
 
+  private boolean isAtleastFiveDotZero() {
+    return getMongoClient()
+            .getDatabase("admin")
+            .runCommand(BsonDocument.parse("{isMaster: 1}"))
+            .get("maxWireVersion", 0)
+        >= 13;
```
Comment (stIncMale) [inline src/integrationTest/java/com/mongodb/kafka/connect/ConnectorValidationIntegrationTest.java:635]: This method is different from `TimeseriesValidation.isAtleastFiveDotZero`. Maybe it is worth marking it public in `TimeseriesValidation` and reusing in tests?

**Review by stIncMale — CHANGES_REQUESTED** [review]

File: src/main/java/com/mongodb/kafka/connect/util/TimeseriesValidation.java [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+        timeSeriesOptions.granularity(TimeSeriesGranularity.SECONDS);
+        break;
+      case MINUTES:
+        timeSeriesOptions.granularity(TimeSeriesGranularity.MINUTES);
+        break;
+      case HOURS:
+        timeSeriesOptions.granularity(TimeSeriesGranularity.HOURS);
+        break;
+      default:
+        // Do nothing
```
[the remark this answers is not in the document]
Comment (rozza) [reply src/main/java/com/mongodb/kafka/connect/util/TimeseriesValidation.java:226]: The default value is TimeSeriesGranularity.DEFAULT - where we should do nothing and not set an option.

File: src/main/java/com/mongodb/kafka/connect/util/TimeseriesValidation.java (hunk 1 above)
Comment (rozza) [reply src/main/java/com/mongodb/kafka/connect/util/TimeseriesValidation.java:124]: So this is a check to see if the collection actually exists.  Unfortunately the documentation doesn't reflect reality.

```
MongoDB Enterprise > show databases
admin   0.000GB
config  0.000GB
local   0.000GB
MongoDB Enterprise > use test
switched to db test
MongoDB Enterprise > var collStats = db.runCommand({collStats: "normal1"});
MongoDB Enterprise > collStats
{
	"ns" : "test.normal1",
	"size" : 0,
	"count" : 0,
	"storageSize" : 0,
	"totalSize" : 0,
	"nindexes" : 0,
	"totalIndexSize" : 0,
	"indexDetails" : {
		
	},
	"indexSizes" : {
		
	},
	"scaleFactor" : 1,
	"ok" : 1
}
```

There is no error on 4.2, 4.4 or the latest rc. So I can use the same command to determine if the collection exists.

File: src/main/java/com/mongodb/kafka/connect/util/TimeseriesValidation.java [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+        return;
+      }
+      if (collStats.getInteger("nindexes") > 0) {
+        throw new ConfigException(
+            format(
+                "A collection already exists for: `%s` that is not a timeseries collection.",
+                namespace.getFullName()));
+      }
+      createCollection(mongoClient, namespace, createCollectionOptions(config));
+    } catch (MongoException e) {
```
[the remark this answers is not in the document]
Comment (rozza) [reply src/main/java/com/mongodb/kafka/connect/util/TimeseriesValidation.java:131]: getCollStats could also error - so it handles both error scenarios.

File: src/main/java/com/mongodb/kafka/connect/util/TimeseriesValidation.java [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+    return getCollStats(mongoClient, namespace).containsKey(TIMESERIES);
+  }
+
+  private static boolean isAtleastFiveDotZero(final MongoClient mongoClient) {
+    try {
+      return mongoClient
+              .getDatabase("admin")
+              .runCommand(BsonDocument.parse("{hello: 1}"))
+              .get("maxWireVersion", 0)
+          >= 13;
```
[the remark this answers is not in the document]
Comment (rozza) [reply src/main/java/com/mongodb/kafka/connect/util/TimeseriesValidation.java:266]: I grabbed that from the Java driver. See: JAVA-4185

File: src/main/java/com/mongodb/kafka/connect/sink/MongoSinkTopicConfig.java [hunk 6]
```diff
@@ -341,6 +352,38 @@ public String value() {
       "The class name of the CDC handler to use for processing";
   private static final String CHANGE_DATA_CAPTURE_HANDLER_DEFAULT = EMPTY_STRING;
 
+  // Timeseries
+  public static final String TIMESERIES_TIMEFIELD_CONFIG = "timeseries.timefield";
+  private static final String TIMESERIES_TIMEFIELD_DISPLAY = "The field used for time";
+  private static final String TIMESERIES_TIMEFIELD_DOC =
+      "Name of the top level field used for time. "
+          + "Note: Inserted documents must have this field.";
```
[the remark this answers is not in the document]
Comment (rozza) [reply src/main/java/com/mongodb/kafka/connect/sink/MongoSinkTopicConfig.java:360]: Great catch :+1:

File: src/main/java/com/mongodb/kafka/connect/util/TimeseriesValidation.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (rozza) [reply src/main/java/com/mongodb/kafka/connect/util/TimeseriesValidation.java:266]: Heres a definitive list: https://github.com/mongodb/specifications/blob/master/source/wireversion-featurelist.rst

File: src/main/java/com/mongodb/kafka/connect/sink/MongoSinkTask.java [hunk 7]
```diff
@@ -185,13 +190,23 @@ private MongoClient getMongoClient() {
     return mongoClient;
   }
 
+  private void checkTimeseries(final MongoNamespace namespace, final MongoSinkTopicConfig config) {
+    if (!checkedTimeseriesNamespaces.contains(namespace)) {
+      if (config.isTimeseries()) {
+        validateCollection(getMongoClient(), namespace, config);
```
[the remark this answers is not in the document]
Comment (rozza) [reply src/main/java/com/mongodb/kafka/connect/sink/MongoSinkTask.java:196]: Good question. This is due to some of the flexibility (complexity) of the Kafka connector.

Depending on how the user has configured their connector, we may not know if a message from Kafka is to be stored in a timeseries collection at start up. Users can store messages from multiple topics and override configurations based on each topic. Further more they can also set the namespace to sink to based on the data within a message from Kafka. 

This means a runtime check is also required.

File: src/integrationTest/java/com/mongodb/kafka/connect/ConnectorValidationIntegrationTest.java (hunk 2 above)
Comment (rozza) [reply src/integrationTest/java/com/mongodb/kafka/connect/ConnectorValidationIntegrationTest.java:635]: Agreed

File: src/main/java/com/mongodb/kafka/connect/util/TimeseriesValidation.java [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+        new TimeSeriesOptions(config.getString(TIMESERIES_TIMEFIELD_CONFIG).trim());
+
+    String metaField = config.getString(TIMESERIES_METAFIELD_CONFIG).trim();
+    if (!metaField.isEmpty()) {
+      timeSeriesOptions.metaField(metaField);
+    }
+
+    MongoSinkTopicConfig.TimeSeriesGranularity granularity =
+        MongoSinkTopicConfig.TimeSeriesGranularity.valueOf(
+            config.getString(TIMESERIES_GRANULARITY_CONFIG).trim().toUpperCase());
```
[the remark this answers is not in the document]
Comment (rozza) [reply src/main/java/com/mongodb/kafka/connect/util/TimeseriesValidation.java:214]: Ack

File: src/main/java/com/mongodb/kafka/connect/sink/MongoSinkTopicConfig.java [hunk 9]
```diff
@@ -341,6 +352,38 @@ public String value() {
...
+  private static final String TIMESERIES_TIMEFIELD_DEFAULT = EMPTY_STRING;
+
+  public static final String TIMESERIES_METAFIELD_CONFIG = "timeseries.metafield";
+  private static final String TIMESERIES_METAFIELD_DISPLAY = "The field describing the series";
+  private static final String TIMESERIES_METAFIELD_DOC =
+      "Name of the top-level field describing the series. "
+          + "Used to group related data and may be any BSON type except array. May not be the same as `timefield` or `_id` value.";
```
[the remark this answers is not in the document]
Comment (rozza) [reply src/main/java/com/mongodb/kafka/connect/sink/MongoSinkTopicConfig.java:367]: Done

File: src/main/java/com/mongodb/kafka/connect/util/TimeseriesValidation.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (rozza) [reply src/main/java/com/mongodb/kafka/connect/util/TimeseriesValidation.java:226]: Updated to remove the need for the copied `TimeSeriesGranularity` and added an error for the default case.

File: src/main/java/com/mongodb/kafka/connect/util/TimeseriesValidation.java (hunk 1 above)
Comment (rozza) [reply src/main/java/com/mongodb/kafka/connect/util/TimeseriesValidation.java:124]: Looks like it errored in 3.6 and previous. As this code is already guarded against pre 5.0 so it won't error. 
Added DOCSP-17199 for the docs to be updated.

File: src/main/java/com/mongodb/kafka/connect/sink/MongoSinkTopicConfig.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (rozza) [reply src/main/java/com/mongodb/kafka/connect/sink/MongoSinkTopicConfig.java]: Done

File: src/integrationTest/java/com/mongodb/kafka/connect/ConnectorValidationIntegrationTest.java (hunk 2 above)
Comment (rozza) [reply src/integrationTest/java/com/mongodb/kafka/connect/ConnectorValidationIntegrationTest.java]: Done

File: src/main/java/com/mongodb/kafka/connect/util/TimeseriesValidation.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (stIncMale) [reply src/main/java/com/mongodb/kafka/connect/util/TimeseriesValidation.java]: Thank you for the link 👏

File: src/main/java/com/mongodb/kafka/connect/util/MongoClientHelper.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+  private MongoClientHelper() {}
+
+  public static boolean isAtleastFiveDotZero(final MongoClient mongoClient) {
+    try {
+      return mongoClient
+              .getDatabase("admin")
+              .runCommand(BsonDocument.parse("{hello: 1}"))
+              .get("maxWireVersion", 0)
+          >= 13;
+    } catch (Exception e) {
```
Comment (stIncMale) [inline src/main/java/com/mongodb/kafka/connect/util/MongoClientHelper.java:34]: [optional]

The code within `try` does not throw checked exceptions, so I propose narrowing the `catch` block to unchecked ones (`RuntimeException`).

File: src/main/java/com/mongodb/kafka/connect/util/TimeseriesValidation.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (stIncMale) [reply src/main/java/com/mongodb/kafka/connect/util/TimeseriesValidation.java:121]: 👍

File: src/main/java/com/mongodb/kafka/connect/util/TimeseriesValidation.java [hunk 11]
```diff
@@ -209,21 +199,22 @@ private static CreateCollectionOptions createCollectionOptions(
...
-        break;
-      case HOURS:
-        timeSeriesOptions.granularity(TimeSeriesGranularity.HOURS);
-        break;
-      default:
-        // Do nothing
+    String granularityValue =
+        config.getString(TIMESERIES_GRANULARITY_CONFIG).trim().toUpperCase(Locale.ROOT);
+    if (!granularityValue.isEmpty()) {
+      switch (TimeSeriesGranularity.valueOf(granularityValue)) {
```
Comment (stIncMale) [inline src/main/java/com/mongodb/kafka/connect/util/TimeseriesValidation.java:205]: This method throws an `IllegalArgumentException` if `granularityValue` cannot be converted to a `TimeSeriesGranularity`. Should it be wrapped into a `ConfigException`?

**Review by stIncMale — CHANGES_REQUESTED** [review]

File: src/main/java/com/mongodb/kafka/connect/util/TimeseriesValidation.java (hunk 1 above)
Comment (stIncMale) [reply src/main/java/com/mongodb/kafka/connect/util/TimeseriesValidation.java]: @rozza, thank you for the detailed explanation and for filing DOCSP-17199!

File: src/main/java/com/mongodb/kafka/connect/util/TimeseriesValidation.java (hunk 11 above)
Comment (rozza) [reply src/main/java/com/mongodb/kafka/connect/util/TimeseriesValidation.java:205]: The validators in the MongoSinkTopicConfig ensure that its either an empty string or can be converted.

File: src/main/java/com/mongodb/kafka/connect/util/MongoClientHelper.java (hunk 10 above)
Comment (rozza) [reply src/main/java/com/mongodb/kafka/connect/util/MongoClientHelper.java:34]: Done

File: src/main/java/com/mongodb/kafka/connect/util/TimeseriesValidation.java (hunk 11 above)
Comment (rozza) [reply src/main/java/com/mongodb/kafka/connect/util/TimeseriesValidation.java:205]: Done

Comment (stIncMale) [reply src/main/java/com/mongodb/kafka/connect/util/TimeseriesValidation.java]: > The validators in the MongoSinkTopicConfig ensure that its either an empty string or can be converted.

Sorry, I forgot there is `Validators.emptyString().or(Validators.EnumValidatorAndRecommender.in(TimeSeriesGranularity.values()))` in `MongoSinkTopicConfig`. If I realized that, I would have not left the comment.

**Review by stIncMale — APPROVED** [review]

**Pull request merged**